### ISSUE TITLE
Unfolded model `Update` gives "TypeError: unsupported operand type(s) for +: 'DynamicJaxprTracer' and 'Diff'"

### ISSUE THREAD

**Issue by derifatives** [genjax-community/genjax#459]
class O(NamedTuple):
  o: Any

def no_diffs(p):
    return jtu.tree_map(lambda x: diff(x, NoChange), p)

@genjax.gen
def f(x_o, y_o):
    o = genjax.tfp_normal(x_o.o + y_o.o, 1.) @ 'o'
    return O(o)

@genjax.gen
def kernel(state, ys_o):
    i, x_o = state
    new_o = f.inline(x_o, jtu.tree_map(lambda r: r[i], ys_o))
    return (i+1, new_o)

@genjax.gen
def unfolded(x_o, ys_o):
    chain = genjax.Unfold(kernel, max_length=10)
    _, xs = chain(9, (0, x_o), ys_o) @ 'c'
    return xs
                 
key, simulate_key, update_key = jax.random.split(key, 3)
model_args = (O(1.), O(jnp.float32(jnp.arange(10))))
tr = unfolded.simulate(simulate_key, model_args)

update_chm = genjax.choice_map(
    {'c': genjax.index_choice_map(
        [3],
        genjax.choice_map({'o': jnp.array([8.])}))})

(_, _, new_tr, _) = unfolded.update(
    update_key, tr, update_chm, no_diffs(model_args))
jax.block_until_ready(new_tr)

**Comment by derifatives** [comment]
Slightly simplified version that still fails:
```
class O(NamedTuple):
  o: Any

def no_diffs(p):
    return jtu.tree_map(lambda x: diff(x, NoChange), p)

@genjax.gen
def f(x_o, y):
    o = genjax.tfp_normal(x_o.o + y, 1.) @ 'o'
    return O(o)

@genjax.gen
def kernel(state, ys):
    i, x_o = state
    new_o = f(x_o, jtu.tree_map(lambda x: x[i], ys)) @ 'f'
    return (i+1, new_o)

@genjax.gen
def unfolded(x_o, ys):
    chain = genjax.Unfold(kernel, max_length=10)
    _, xs = chain(9, (0, x_o), ys) @ 'c'
    return xs
                 
key, simulate_key, update_key = jax.random.split(key, 3)
model_args = (O(1.), jnp.float32(jnp.arange(10)))
tr = unfolded.simulate(simulate_key, model_args)

update_chm = genjax.choice_map(
    {'c': genjax.index_choice_map(
        [3],
        genjax.choice_map({('f', 'o'): jnp.array([8.])}))})

(_, _, new_tr, _) = unfolded.update(
    update_key, tr, update_chm, no_diffs(model_args))
jax.block_until_ready(new_tr)
```

One *super* interesting there is the current definition of `kernel`. Note that `ys` is just a jnp.array here, but if I replace this line with:
```
    new_o = f(x_o, ys[i])) @ 'f'
```
I get a different error:
```
TypeError: 'Diff' object is not subscriptable
```

**Comment by derifatives** [comment]
I think this is fixed in the `fix_slam_451` branch.

**Comment by femtomc** [comment]
Closed by #460.
